### PR TITLE
Oidc JWE FAT delivery #2

### LINKED ISSUE #17035: Feature Test Summary for OIDC JWE (RP) support
### List of FAT projects affected

* **com.ibm.ws.security.oidc.client_fat.1** - com.ibm.ws.security.oidc.client_fat was split (into com.ibm.ws.security.oidc.client_fat.1 and com.ibm.ws.security.oidc.client_fat.2) as it was getting to large and I needed to add tests for this feature.   The project was updated to use shared config sections.
* **com.ibm.ws.security.oidc.client_fat.2** - Part of the com.ibm.ws.security.oidc.client_fat project split.  New tests were added to test the new Signature Algorithms, processing of JWE's from the token endpoint, JWS and JWE reponses from the userinfo endpoint (RP)
* **com.ibm.ws.security.oidc.client_fat.jaxrs.config** - New tests were added to test the new Signature Algorithms, JWE's, WS and JWE reponses from the userinfo endpoint in the propagation path(RS)
* **com.ibm.ws.security.social_fat** - New common tests were added to test the new Signature Algorithms, processing of JWE's from the token endpoint, JWS and JWE reponses from the userinfo endpoint.
* **com.ibm.ws.security.social_fat.LibertyOP** -  New tests classes were added to run the new common social tests with a LibertyOP as the provider.
* **com.ibm.ws.security.oauth.oidc_fat.common** - new tooling and configs were added to support the new Signature Algorithms, processing of JWE's, test application stubs for the token and userinfo endpoint (needed since our OP does not generate JWE's)
* **com.ibm.ws.security.fat.common.jwt** - updated shared config sections (ie: common keystore definitions, ...)

### Test strategy

This feature is adding support for the following:
- support additional algorithms for signing of JWT token
- support JWE's as access tokens
- support responses from the userinfo endpoint in the form of json data, JWS' or JWE's (we supported json only before this)

Things to note:
- This feature is adding support in our our RP, RS and Social **clients**.  The Liberty OP does not currently support the additional signature algorithms, JWE's or any other format response than json from the userinfo endpoint.
- Since our OP does not currently support what our client needs to test, we've had to create some test apps that we use in place of the OP's token and userinfo endpoints.
- Both test the token endpoint and userinfo test apps will be called by our test cases and will store/save a token created for each specific test case.  When the product runtime calls these endpoints (instead of the endpoints that are part of the product), the endpoints will return the token previously saved by the test case.  (The client configs specify the actual url that should be used for each endpoint - this is what allows us to replace the real endpoint with our test version)
- These test endpoints allow us to test that our RP, RS and Social clients can handle the new tokens using the normal end to end product flow.

**New Signature algorithms**:
- **RP**
  - Test that an access token created using each of the supported signature algorithms can be consumed by our RP (oidcclient) that has the same configured signature algorithm.  
  - Test that an access token created using each of the supported signature algorithms will NOT result in access to a protected app if our RP (oidcclient) is configured with a different signature algorithm
  - Test that an access token created using each of the supported signature algorithms will NOT result in access to a protected app if our RP (oidcclient) is configured using the same signature algorithm, but, a different key
- **RS** (token propagation)
  - Test that an access token created using each of the supported signature algorithms can be consumed by our RS (oidcclient - with propagation) that has the same configured signature algorithm.  
  - Test that an access token created using each of the supported signature algorithms will NOT result in access to a protected app if our RS (oidcclient - with propagation) is configured with a different signature algorithm
  - Test that an access token created using each of the supported signature algorithms will NOT result in access to a protected app if our RS (oidcclient) is configured using the same signature algorithm, but, a different key
- **Social Client**
  - Test that an access token created using each of the supported signature algorithms can be consumed by our Social Client that has the same configured signature algorithm.  
  - Test that an access token created using each of the supported signature algorithms will NOT result in access to a protected app if our Social Client is configured with a different signature algorithm
  - Test that an access token created using each of the supported signature algorithms will NOT result in access to a protected app if our Social Client is configured using the same signature algorithm, but, a different key

**JWE support**:
- **RP**
  - Test that an access token created using each of the supported encryption algorithms can be consumed by our RP (oidcclient) that has the same configured encryption settings.  (This includes testing with access_tokens configured to use the same alg for signing and encrypting as well as using different algs for signing vs encrypting - clients match the token they're given)
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our RP (oidcclient) is configured with a different encryption settings
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our RP (oidcclient) is configured using the same encryption settings, but, a different key
  - Test that an access token that is NOT encrypted will NOT result in access to a protected app if our RP (oidcclient) is configured using each of the supported encryption algorithms
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our RP (oidcclient) is not configured to use encryption.
  - Test that an access token with an invalid format - wrong number of parts, invalid value within each part (ie: just a simple string), simple JSON payload, ... will NOT result in access to a protected app if our RP (oidcclient) is configured to expect a JWE.
- **RS** (token propagation)
  - Test that an access token created using each of the supported encryption algorithms can be consumed by our RS (oidcclient) that has the same configured encryption settings.  (This includes testing with access_tokens configured to use the same alg for signing and encrypting as well as using different algs for signing vs encrypting - clients match the token they're given)
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our RS (oidcclient) is configured with a different encryption settings
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our RS (oidcclient) is configured using the same encryption settings, but, a different key
  - Test that an access token that is NOT encrypted will NOT result in access to a protected app if our RS (oidcclient) is configured using each of the supported encryption algorithms
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our RS (oidcclient) is not configured to use encryption.
  - Test that an access token with an invalid format - wrong number of parts, invalid value within each part (ie: just a simple string), simple JSON payload, ... will NOT result in access to a protected app if our RS (oidcclient) is configured to expect a JWE.
- **Social Client**
  - Test that an access token created using each of the supported encryption algorithms can be consumed by our Social Client that has the same configured encryption settings.  (This includes testing with access_tokens configured to use the same alg for signing and encrypting as well as using different algs for signing vs encrypting - clients match the token they're given)
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our Social Client is configured with a different encryption settings
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our Social Client is configured using the same encryption settings, but, a different key
  - Test that an access token that is NOT encrypted will NOT result in access to a protected app if our Social Client is configured using each of the supported encryption algorithms
  - Test that an access token created using each of the supported encryption algorithms will NOT result in access to a protected app if our Social Client is not configured to use encryption.
  - Test that an access token with an invalid format - wrong number of parts, invalid value within each part (ie: just a simple string), simple JSON payload, ... will NOT result in access to a protected app if our Social Client is configured to expect a JWE.
  - Test that an access token with invalid content in the JWE header will NOT result in access to a protected app.

**userinfo:**
- Test with userinfo returning a JWS or JWE that conforms to the RP (oidcclient), RS (oidcclient - with propagation) or the Social Client config, ensure that our runtime can comsume it propertly - extra claims from the userinfo reponse are added to the subject
- Test with userinfo returning a JWS or JWE that is not valid for the RP (oidcclient), RS (oidcclient - with propagation) or the Social Client  config 
  - when the validation type is introspect, introspect is to be used to validate the token (the token will validate), the runtime also calls userinfo to add any extra claims - this will fail because the token doesn't meet what the config requires - the overall validate request will succeed, but none of the extra claims will be added to the subject
  - when the validation type is userinfo, the validation request will fail as the token doesn't meet what the config requires.

List of tests added (308 tests): 
[14646TestList.zip](https://github.com/OpenLiberty/open-liberty/files/7289823/14646TestList.zip)

**Positive tests - Validation of tests include**:
1.  200 status code
2. - Correct application output content (valid app output in the response) - The token built by/for our token and userinfo test endpoints adds additional claims that the tests search for to prove that we have in fact consumed the tokens of the format that we're testing.
3. When testing userinfo ensure that additional claims from the response from userinfo are included in the subject built and logged by the test app
3. - no FFDC's or error messages in log during test case execution (part of the overall framework)

**Negative tests - Validation of tests include**:
1.  401 status code
2. Tests locate error messages appropriate for the test cases expected failure in the proper server log. 
3. Tests receive the expected and appropriate ffdc's 

There are NO manual tests required to test this feature

## Confidence Level
5 - All possible combinations are being tested via automation.
In the future, we will have to update the tests to use our OP once it is updated to support JWE's

### PR DESCRIPTION
**Pull request by c00crane**
Deliver FAT tests to verify support for JWT Tokens by the OIDC RP

### CODE REVIEW

**Review by ayoho — COMMENTED** [review]
Changes look fine

File: dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops [hunk 1]
```diff
@@ -382,3 +382,7 @@ JWE_REQUIRED_BUT_TOKEN_NOT_JWE.useraction=To accept JWS tokens instead of JWE to
 NESTED_JWS_REQUIRED_BUT_NOT_FOUND=CWWKS6065E: The payload of the JSON Web Encryption (JWE) token that is included in the request is not a JSON Web Signature (JWS) token.
 NESTED_JWS_REQUIRED_BUT_NOT_FOUND.explanation=The payload of a JWE token must be a valid JSON Web Token (JWT) in JWS format.
 NESTED_JWS_REQUIRED_BUT_NOT_FOUND.useraction=Ensure that the payload of the JWE token is a valid JWT in JWS format.
+
+JWE_DECRYPTION_KEY_MISSING=CWWKS6066E: The JSON Web Encryption (JWE) token cannot be decrypted because a decryption key cannot be found. The {0} server configuration attribute is set to [{1}].
+JWE_DECRYPTION_KEY_MISSING.explanation=A decryption key is required to decrypt the JWE token. The decryption key must be present in the key store that is configured for the JWT consumer, or the key must be specified in MicroProfile Config properties.
```
Comment (helyarp) [inline dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops]: - Change "key store" to "keystore" (1 word)
- Put "the" before "MicroProfile Config properties"

The decryption key must be present in the keystore that is configured for the JWT consumer, or the key must be specified in the MicroProfile Config properties.

File: dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops [hunk 2]
```diff
@@ -382,3 +382,7 @@ JWE_REQUIRED_BUT_TOKEN_NOT_JWE.useraction=To accept JWS tokens instead of JWE to
...
+JWE_DECRYPTION_KEY_MISSING.useraction=Verify that the key alias that is specified in the message exists in the key store that is configured for the JWT consumer. If the alias is not configured, the decryption key must be specified in MicroProfile Config properties.
```
Comment (helyarp) [inline dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops]: - Same as previous comment: Use "keystore" and "the MicroProfile Config properties."
- Unless the wording "the keystore that is configured for the JWT consumer" is needed for clarity, shorten the wording to "the JWT consumer keystore."

**Review by helyarp — COMMENTED** [review]
Suggested changes.

**Comment by skasund** [comment]
L2 message review completed

**Comment by c00crane** [comment]
#run_libby_bot

**Comment by c00crane** [comment]
#run-libby-bot

File: dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops (hunk 1 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops]: Adam made the changes under 17511 - my pull had drug his product changes along for the ride.
We've now separated things...

File: dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops (hunk 2 above)
Comment (c00crane) [reply dev/com.ibm.ws.security.jwt/resources/com/ibm/ws/security/jwt/internal/resources/JWTMessages.nlsprops]: Adam made the changes under 17511 - my pull had drug his product changes along for the ride.
We've now separated things...

**Comment by c00crane** [comment]
#build

**Review by ayoho — APPROVED** [review]

**Comment by c00crane** [comment]
#run-libby-bot
#build

**Comment by c00crane** [comment]
#libby

**Pull request merged**